### ISSUE TITLE
how to use CallbackQueryOriginHandler ?

### ISSUE THREAD

**Issue by converge** [nickoala/telepot#175]
Hi,

I'm trying to get an user action (wanna know what button user clicked), is **CallbackQueryOriginHandler** the right way ?

I found it on Quizz example, but it's kind difficult to understand how it works, can anyone help me ?

What I'm trying to do is:  
1. give the user 3 options, this options are dynamic (that's why I need to know what he clicked)
2. use the user option as argument for the next action.

**Comment by nickoala** [comment]
Let me explain what is happening in the quiz example:

```python
bot = telepot.DelegatorBot(TOKEN, [
    pave_event_space()(
        per_chat_id(), create_open, QuizStarter, timeout=3),
    pave_event_space()(
        per_callback_query_origin(), create_open, Quizzer, timeout=10),
])
```
1. As a user sends a message to the bot, `per_chat_id()` would spawn a `QuizStarter`, which sends back a message containing a START button.
2. As the START button is pressed, a callback query is generated. That callback query causes a `Quizzer` to be spawned, which changes the original message to a multiple-choice math question. Note that the `Quizzer` never sends a new message. It only edits the original message over and over.
3. The user picks a choice. That generates another callback query, which causes the existing `Quizzer` to record the answer and change the message to another multiple-choice math question. This callback query has the same *origin* as the last callback query, because they both come from the same message (which is being edited by `Quizzer`over and over).
4. Because all callback queries share the same origin, `per_callback_query_origin()` would "group" them into one `Quizzer` object. No new `Quizzer` object will be spawned for subsequent callback queries.

In this example, `QuizStarter` and `Quizzer` are two "threads". The former creates the message which initially contains only the START button. The latter recycles that message to show questions.

From your description, it seems you want the option-picking to be a part of an ongoing conversation. I think it's more natural to handle chat messages and callback queries in one "thread". The **Lover** example is more relevant in this case:
```python
bot = telepot.DelegatorBot(TOKEN, [
    include_callback_query_chat_id(
        pave_event_space())(
            per_chat_id(types=['private']), create_open, Lover, timeout=10),
])
```
Simply put, `include_callback_query_chat_id()` causes callback queries originated from a chat to be "grouped" into the same `Lover`. Just define `on_chat_message()` and `on_callback_query()` accordingly.

The Quiz example could be implemented this way, too. The difference is, because `Quizzer` is `per_callback_query_origin` (that is, per original message), it's possible that a user can spawn more than one `Quizzer` (that is, more than one line of questioning). If you use `include_callback_query_chat_id()` as in the Lover example, a user can only spawn one `Lover`.

Hope that helps. Good luck.

**Comment by converge** [comment]
@nickoala thanks, that helped me.

the main problem I have is to understand how can I get the user options. It will be a sequence of options, like:

select a brand:
1. BMW (my choice)
2. Mercedes

choose a BMW model:
1. X1
2. X2 (my choice)

choose a color:
1. black (my choice)
2. white

In the end the bot prints: You selected the "black BMW X2"

is there a way to do it ? Im in the very first steps trying to connect on my db. 

my code: https://gist.github.com/converge/7b8dce953dc10b6b53c89d73e093580c

**Comment by nickoala** [comment]
Well, here is an outline:

1. You use `InlineKeyboardMarkup` and `InlineKeyboardButton` to compose the inline keyboard.
2. Override `on_chat_message()` to react to a user's initial message
3. Override `on_callback_query()` to react to user pressing a button. From the `data` field, you know which option he has picked. Then, you may edit the original message to show the next inline keyboard.

I am sorry I can't write the code for you. The Lover example should be a good starting point. The key is `on_chat_message()` and `on_callback_query()`.
